renku/core/incubation/database.py

```diff
@@ -0,0 +1,466 @@
+        while self._objects_to_commit:
+            oid, object = self._objects_to_commit.popitem()
+
+            if not object._p_changed and object._p_serial != NEW:
+                continue
+
+            object_type = type(object).__name__
+            if object_type in self._root_type_names:
+                self._create_root_type(self.root, object_type)
+                oid = oid.decode("ascii")
+                root_object = self.root[object_type]
+                root_object[oid] = object
+                root_object._p_changed = 1
+
+            self._store_object(object)
+
+    def _store_object(self, object: Persistent):
+        oid = object._p_oid
+
+        data = self._writer.serialize(object)
+
+        self._storage.store(oid=oid, data=data)
+        self._cache[oid] = object
+
+        self._cache.update_object_size_estimation(oid, 1)
+        object._p_estimated_size = 1
+
+        object._p_changed = 0  # NOTE: transition from changed to up-to-date
+        object._p_serial = PERSISTED
+
+    def new_ghost(self, oid: bytes, object: Persistent):
+        """Create a new ghost object."""
+        self._cache.new_ghost(oid, object)
+
+    @classmethod
+    def from_path(cls, path: Union[str, Path]) -> "Database":
+        """Return a Database."""
+        return cls(path)
+
+    def replace(self, object: Persistent):
+        """Remove an object by creating a new oid for it."""
+        try:
+            del self._cache[object._p_oid]
+        except KeyError:
+            pass
+        else:
+            type_name = type(object).__name__
+            if type_name in self._root_type_names:
+                root_object = self.root[type_name]
+                del root_object[object._p_oid.decode("ascii")]
+                root_object._p_changed = 1
+
+        object._p_oid = None
+        self.add(object)
+
+    def readCurrent(self, object):
+        """We don't use this method but some Persistent logic require its existence."""
+        assert object._p_jar is self
+        assert object._p_oid is not None
+
+
+class Storage:
+    """Store Persistent objects on the disk."""
+
+    def __init__(self, path: Union[Path, str]):
+        self.path = Path(path)
+        self.path.mkdir(parents=True, exist_ok=True)
+
+    def store(self, oid: bytes, data: Union[Dict, List]):
+        """Store object."""
+        assert isinstance(oid, bytes)
+
+        filename = oid.decode("ascii").lower()
+        compressed = len(filename) >= MIN_COMPRESSED_FILENAME_LENGTH
+
+        if compressed:
+            path = self.path / filename[0:2] / filename[2:4] / filename
+            path.parent.mkdir(parents=True, exist_ok=True)
+            open_func = open  # TODO: Change this to gzip.open for the final version
+        else:
+            path = self.path / filename
+            open_func = open
+
+        with open_func(path, "w") as file:
+            json.dump(data, file, ensure_ascii=False, sort_keys=True, indent=2)
```

Comment:
It'd be nice if `Storage` just had two method stubs
```
def load(path):
    raise NotImplementedException()

def dump(data, path):
    raise NotImplementedException()
```

And then we can have a child class `JsonStorage`, `GZipStorage` (Which could just wrap any storage), `MsgPackStorage` that implement these functions. Then we can easily do comparisons and it's a bit less tighly coupled.

Reply:
I've changed it to be like this with a difference that it accepts filename instead of full path.